### User Engagement Analysis

#### Challenge
The product team at Showwcase wants to better understand "user engagement" on the platform in the last month
(October 2019 in this exercise). You are free to define engagement, find insights, and question any other metrics we
can use to understand our approach better. Your role is to present your findings based on the data. Note - The
sample dataset may have data integrity issues; if so, you should base your analysis accordingly.

#### Dataset
Dataset consist of records of each session a user completes and has following columns:<br>
● session_id → unique identification number for each session <br>
● customer_id → unique identification number for each customer <br>
● login_date → the date the user logged into Showwcase for that session<br>
● projects_added → boolean if at least one project was added <br>
● likes_given → boolean if at least one like was given <br>
● comment_given → boolean if at least one comment was given <br>
● inactive_status → boolean if the user was inactive or idle <br>
● bugs_occurerd → boolean if a user ran into at least one bug <br>
● session_projects_added → number of projects added <br>
● session_likes_given → number of likes given <br>
● session_comments_given → number of comments given <br>
● inactive_duration → number of seconds a user was inactive or idle <br>
● bugs_in_session → number of bugs a user ran into <br>
● session_duration → number of seconds a user was logged for that session <br>

My Approach :
In real life, we analyze and try to get insights from the data as we go. This notebook follows the same approach. While analyzing the data, insights are captured and noted.

In [ ]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('showwcase_sessions.csv')

In [3]:
data.head()

session_id  customer_id login_date projects_added likes_given  \
0    624205.0      80746.0   10/30/19          False        True   
1    624241.0      24520.0   10/30/19           True        True   
2    111002.0      32047.0   10/30/19           True        True   
3    545113.0      23404.0   10/30/19           True        True   
4    750269.0      40235.0   10/30/19           True        True   

  comment_given inactive_status bug_occured  session_projects_added  \
0          True            True       False                     0.0   
1          True            True       False                     2.0   
2          True            True       False                     1.0   
3          True           False       False                     1.0   
4         False            True       False                     3.0   

   session_likes_given  session_comments_given  inactive_duration  \
0                 24.0                     3.0             1146.0   
1                  3.0                     5.0              133.0   
2                  5.0                     5.0             1571.0   
3                 10.0                    21.0                0.0   
4                 16.0                     0.0             1405.0   

   bugs_in_session  session_duration  
0              0.0            1564.0  
1              0.0            1766.0  
2              0.0            2230.0  
3              0.0             633.0  
4              0.0            1679.0

In [4]:
data['login_date'] = pd.to_datetime(data['login_date'])

In [5]:
data.head()

session_id  customer_id login_date projects_added likes_given  \
0    624205.0      80746.0 2019-10-30          False        True   
1    624241.0      24520.0 2019-10-30           True        True   
2    111002.0      32047.0 2019-10-30           True        True   
3    545113.0      23404.0 2019-10-30           True        True   
4    750269.0      40235.0 2019-10-30           True        True   

  comment_given inactive_status bug_occured  session_projects_added  \
0          True            True       False                     0.0   
1          True            True       False                     2.0   
2          True            True       False                     1.0   
3          True           False       False                     1.0   
4         False            True       False                     3.0   

   session_likes_given  session_comments_given  inactive_duration  \
0                 24.0                     3.0             1146.0   
1                  3.0                     5.0              133.0   
2                  5.0                     5.0             1571.0   
3                 10.0                    21.0                0.0   
4                 16.0                     0.0             1405.0   

   bugs_in_session  session_duration  
0              0.0            1564.0  
1              0.0            1766.0  
2              0.0            2230.0  
3              0.0             633.0  
4              0.0            1679.0

In [6]:
data = data[:-1]

In [7]:
data.tail()

session_id  customer_id login_date projects_added likes_given  \
296    558332.0      87323.0 2019-10-01           True       False   
297    643880.0      51243.0 2019-10-01           True       False   
298    844518.0      23083.0 2019-10-01           True       False   
299    933954.0      38459.0 2019-10-01           True       False   
300         NaN          NaN        NaT            NaN         NaN   

    comment_given inactive_status bug_occured  session_projects_added  \
296         False            True       False                     2.0   
297         False            True       False                     2.0   
298         False            True       False                     1.0   
299         False           False       False                     1.0   
300           NaN             NaN         NaN                     NaN   

     session_likes_given  session_comments_given  inactive_duration  \
296                  0.0                     0.0               97.0   
297                  0.0                     0.0              906.0   
298                  0.0                     0.0              139.0   
299                  0.0                     0.0                0.0   
300                  NaN                     NaN                NaN   

     bugs_in_session  session_duration  
296              0.0            1692.0  
297              0.0            1990.0  
298              0.0            1113.0  
299              0.0             306.0  
300              NaN               NaN

In [8]:
data = data[:-1]

In [9]:
data.tail()

session_id  customer_id login_date projects_added likes_given  \
295    944212.0      40572.0 2019-10-01           True        True   
296    558332.0      87323.0 2019-10-01           True       False   
297    643880.0      51243.0 2019-10-01           True       False   
298    844518.0      23083.0 2019-10-01           True       False   
299    933954.0      38459.0 2019-10-01           True       False   

    comment_given inactive_status bug_occured  session_projects_added  \
295         False            True       False                     3.0   
296         False            True       False                     2.0   
297         False            True       False                     2.0   
298         False            True       False                     1.0   
299         False           False       False                     1.0   

     session_likes_given  session_comments_given  inactive_duration  \
295                 13.0                     0.0             1174.0   
296                  0.0                     0.0               97.0   
297                  0.0                     0.0              906.0   
298                  0.0                     0.0              139.0   
299                  0.0                     0.0                0.0   

     bugs_in_session  session_duration  
295              0.0            2255.0  
296              0.0            1692.0  
297              0.0            1990.0  
298              0.0            1113.0  
299              0.0             306.0

In [10]:
data.isnull().sum()

session_id                0
customer_id               0
login_date                0
projects_added            0
likes_given               0
comment_given             0
inactive_status           0
bug_occured               0
session_projects_added    0
session_likes_given       1
session_comments_given    0
inactive_duration         0
bugs_in_session           0
session_duration          0
dtype: int64

In [11]:
data['session_likes_given'].fillna(0, inplace = True)

In [12]:
cols = ['session_id', 'customer_id', 'session_projects_added', 'session_likes_given','session_comments_given','inactive_duration','bugs_in_session',
       'session_duration']

In [13]:
data[cols] = data[cols].applymap(np.int64)

In [14]:
data

session_id  customer_id login_date projects_added likes_given  \
0        624205        80746 2019-10-30          False        True   
1        624241        24520 2019-10-30           True        True   
2        111002        32047 2019-10-30           True        True   
3        545113        23404 2019-10-30           True        True   
4        750269        40235 2019-10-30           True        True   
..          ...          ...        ...            ...         ...   
295      944212        40572 2019-10-01           True        True   
296      558332        87323 2019-10-01           True       False   
297      643880        51243 2019-10-01           True       False   
298      844518        23083 2019-10-01           True       False   
299      933954        38459 2019-10-01           True       False   

    comment_given inactive_status bug_occured  session_projects_added  \
0            True            True       False                       0   
1            True            True       False                       2   
2            True            True       False                       1   
3            True           False       False                       1   
4           False            True       False                       3   
..            ...             ...         ...                     ...   
295         False            True       False                       3   
296         False            True       False                       2   
297         False            True       False                       2   
298         False            True       False                       1   
299         False           False       False                       1   

     session_likes_given  session_comments_given  inactive_duration  \
0                     24                       3               1146   
1                      3                       5                133   
2                      5                       5               1571   
3                     10                      21                  0   
4                     16                       0               1405   
..                   ...                     ...                ...   
295                   13                       0               1174   
296                    0                       0                 97   
297                    0                       0                906   
298                    0                       0                139   
299                    0                       0                  0   

     bugs_in_session  session_duration  
0                  0              1564  
1                  0              1766  
2                  0              2230  
3                  0               633  
4                  0              1679  
..               ...               ...  
295                0              2255  
296                0              1692  
297                0              1990  
298                0              1113  
299                0               306  

[300 rows x 14 columns]

In [15]:
data.customer_id.nunique()

48

## Insight 1 : 48 unique customers visited in oct

In [16]:
data.tail()

session_id  customer_id login_date projects_added likes_given  \
295      944212        40572 2019-10-01           True        True   
296      558332        87323 2019-10-01           True       False   
297      643880        51243 2019-10-01           True       False   
298      844518        23083 2019-10-01           True       False   
299      933954        38459 2019-10-01           True       False   

    comment_given inactive_status bug_occured  session_projects_added  \
295         False            True       False                       3   
296         False            True       False                       2   
297         False            True       False                       2   
298         False            True       False                       1   
299         False           False       False                       1   

     session_likes_given  session_comments_given  inactive_duration  \
295                   13                       0               1174   
296                    0                       0                 97   
297                    0                       0                906   
298                    0                       0                139   
299                    0                       0                  0   

     bugs_in_session  session_duration  
295                0              2255  
296                0              1692  
297                0              1990  
298                0              1113  
299                0               306

In [17]:
data.login_date.nunique()

30

In [18]:
data.groupby('login_date')['customer_id'].nunique()

login_date
2019-10-01     9
2019-10-02     5
2019-10-03     7
2019-10-04     7
2019-10-05    11
2019-10-06    11
2019-10-07    10
2019-10-08     8
2019-10-09     8
2019-10-10    11
2019-10-11    11
2019-10-12     7
2019-10-13    10
2019-10-14     3
2019-10-15    13
2019-10-16     4
2019-10-17     5
2019-10-18    10
2019-10-19    11
2019-10-20    11
2019-10-21     9
2019-10-22    11
2019-10-23     4
2019-10-24    11
2019-10-25     6
2019-10-26    14
2019-10-27     7
2019-10-28    10
2019-10-29     7
2019-10-30     9
Name: customer_id, dtype: int64

## Insight 2 : Most customers visited was on '2019-10-15' which was 13

In [19]:
data1 = data.copy()

In [20]:
from datetime import datetime

In [21]:
data1.head()

session_id  customer_id login_date projects_added likes_given  \
0      624205        80746 2019-10-30          False        True   
1      624241        24520 2019-10-30           True        True   
2      111002        32047 2019-10-30           True        True   
3      545113        23404 2019-10-30           True        True   
4      750269        40235 2019-10-30           True        True   

  comment_given inactive_status bug_occured  session_projects_added  \
0          True            True       False                       0   
1          True            True       False                       2   
2          True            True       False                       1   
3          True           False       False                       1   
4         False            True       False                       3   

   session_likes_given  session_comments_given  inactive_duration  \
0                   24                       3               1146   
1                    3                       5                133   
2                    5                       5               1571   
3                   10                      21                  0   
4                   16                       0               1405   

   bugs_in_session  session_duration  
0                0              1564  
1                0              1766  
2                0              2230  
3                0               633  
4                0              1679

In [25]:
data1['weekday'] = data1['login_date'].dt.dayofweek

In [26]:
def map_data(day):
    weekDays = ("Monday","Tuesday","Wednesday","Thursday","Friday","Saturday","Sunday")
    return weekDays[day]

In [27]:
data1['weekday'] = data1['weekday'].map(map_data)

In [28]:
data1.head()

session_id  customer_id login_date projects_added likes_given  \
0      624205        80746 2019-10-30          False        True   
1      624241        24520 2019-10-30           True        True   
2      111002        32047 2019-10-30           True        True   
3      545113        23404 2019-10-30           True        True   
4      750269        40235 2019-10-30           True        True   

  comment_given inactive_status bug_occured  session_projects_added  \
0          True            True       False                       0   
1          True            True       False                       2   
2          True            True       False                       1   
3          True           False       False                       1   
4         False            True       False                       3   

   session_likes_given  session_comments_given  inactive_duration  \
0                   24                       3               1146   
1                    3                       5                133   
2                    5                       5               1571   
3                   10                      21                  0   
4                   16                       0               1405   

   bugs_in_session  session_duration    weekday  
0                0              1564  Wednesday  
1                0              1766  Wednesday  
2                0              2230  Wednesday  
3                0               633  Wednesday  
4                0              1679  Wednesday

In [29]:
data1.groupby('weekday')['customer_id'].count()

weekday
Friday       40
Monday       41
Saturday     55
Sunday       42
Thursday     41
Tuesday      49
Wednesday    32
Name: customer_id, dtype: int64

In [ ]:
data1.to_csv('showwcase_sessions_preprocessed.csv', index = False)

## Insight 3 : Saturdays have gotten in the most number of customers on the platform

In [30]:
data1['projects_added'].value_counts()

True     227
False     73
Name: projects_added, dtype: int64

In [31]:
projtrue = data1.loc[data1['projects_added'] == True]

In [32]:
projtrue.head()

session_id  customer_id login_date projects_added likes_given  \
1      624241        24520 2019-10-30           True        True   
2      111002        32047 2019-10-30           True        True   
3      545113        23404 2019-10-30           True        True   
4      750269        40235 2019-10-30           True        True   
5      744943        73245 2019-10-30           True        True   

  comment_given inactive_status bug_occured  session_projects_added  \
1          True            True       False                       2   
2          True            True       False                       1   
3          True           False       False                       1   
4         False            True       False                       3   
5          True            True        True                       3   

   session_likes_given  session_comments_given  inactive_duration  \
1                    3                       5                133   
2                    5                       5               1571   
3                   10                      21                  0   
4                   16                       0               1405   
5                   27                       5               1746   

   bugs_in_session  session_duration    weekday  
1                0              1766  Wednesday  
2                0              2230  Wednesday  
3                0               633  Wednesday  
4                0              1679  Wednesday  
5                4              1490  Wednesday

In [33]:
projtrue.groupby('weekday')['projects_added'].count()

weekday
Friday       27
Monday       33
Saturday     39
Sunday       31
Thursday     31
Tuesday      39
Wednesday    27
Name: projects_added, dtype: int64

## Insight 4 : Most number of projects were added on Saturdays and Tuesdays

In [34]:
liketrue = data1.loc[data1['likes_given'] == True]

In [35]:
liketrue.groupby('weekday')['likes_given'].count()

weekday
Friday       32
Monday       32
Saturday     39
Sunday       30
Thursday     31
Tuesday      29
Wednesday    20
Name: likes_given, dtype: int64

## Insight 5 : Highest engagement in terms of likes has been on Saturday (Weekends have performed the best)

In [36]:
commenttrue = data1.loc[data1['comment_given'] == True]

In [37]:
commenttrue.groupby('weekday')['comment_given'].count()

weekday
Friday       28
Monday       32
Saturday     44
Sunday       37
Thursday     31
Tuesday      31
Wednesday    22
Name: comment_given, dtype: int64

## Insight 6 : Highest engagement in terms of comments has been on Saturday, much higher than any other day

In [38]:
inactivity = data1.loc[data1['inactive_status'] == True]

In [39]:
inactivity.groupby('customer_id')['inactive_status'].count()

customer_id
10246     2
12407     5
12496     4
12940     4
14354     7
19348     1
20548     1
23083     8
23404    10
23579     2
23596     1
23985     6
24520     2
29356     6
29375    13
32047     4
34574     1
38459    10
39475     2
40235     9
40572    11
43057     3
43086     2
43096     1
43549     3
43658     1
46329     1
51243     5
73245    11
73524     6
76530     1
80746     4
82963     1
87265     5
87323     8
92736     1
93464     4
96254     2
98346     1
98653     3
Name: inactive_status, dtype: int64

## Insight 7 : In the month of Oct, customers 23404.0, 29375.0, 38459.0, 40572.0 & 73245.0 have been flagged as the most inactive during their sessions. This can be addressed in a targeted manner

In [40]:
projtrue.head()

session_id  customer_id login_date projects_added likes_given  \
1      624241        24520 2019-10-30           True        True   
2      111002        32047 2019-10-30           True        True   
3      545113        23404 2019-10-30           True        True   
4      750269        40235 2019-10-30           True        True   
5      744943        73245 2019-10-30           True        True   

  comment_given inactive_status bug_occured  session_projects_added  \
1          True            True       False                       2   
2          True            True       False                       1   
3          True           False       False                       1   
4         False            True       False                       3   
5          True            True        True                       3   

   session_likes_given  session_comments_given  inactive_duration  \
1                    3                       5                133   
2                    5                       5               1571   
3                   10                      21                  0   
4                   16                       0               1405   
5                   27                       5               1746   

   bugs_in_session  session_duration    weekday  
1                0              1766  Wednesday  
2                0              2230  Wednesday  
3                0               633  Wednesday  
4                0              1679  Wednesday  
5                4              1490  Wednesday

In [41]:
projtrue.loc[projtrue['bug_occured'] == True].shape

(88, 15)

In [42]:
projtrue.shape

(227, 15)

## Insight 8 : 69% of the projects added had bugs occuring in it at some point

In [43]:
projtrue.groupby('customer_id')['session_projects_added'].sum().sort_values(ascending=False).head(10)

customer_id
87323    38
14354    31
40572    27
23404    25
38459    25
51243    24
73245    24
87265    24
73524    23
40235    23
Name: session_projects_added, dtype: int64

## Insight 9 : Customer no. 87323 added most number of projects in the month of October

In [44]:
data1.groupby('login_date')['session_likes_given'].sum().sort_values(ascending=False)

login_date
2019-10-26    247
2019-10-05    234
2019-10-07    179
2019-10-21    163
2019-10-22    163
2019-10-11    155
2019-10-28    153
2019-10-06    146
2019-10-04    129
2019-10-19    127
2019-10-20    124
2019-10-30    110
2019-10-18    103
2019-10-08    103
2019-10-13    100
2019-10-27     98
2019-10-24     96
2019-10-03     88
2019-10-15     82
2019-10-29     79
2019-10-25     75
2019-10-10     74
2019-10-09     73
2019-10-17     66
2019-10-23     49
2019-10-12     46
2019-10-16     32
2019-10-02     20
2019-10-01     13
2019-10-14      0
Name: session_likes_given, dtype: int64

In [45]:
data1.groupby('customer_id')['session_likes_given'].sum().sort_values(ascending=False)

customer_id
87323    230
23404    207
29375    189
38459    187
73245    183
40572    179
51243    154
14354    150
87265    149
40235    119
23985    115
23083    112
12940     89
29356     87
12407     82
23579     71
73524     65
43057     64
34574     61
96254     59
12496     53
10246     53
43549     52
39475     47
43086     45
93464     44
24520     29
98653     27
29345     27
32047     25
20548     25
68754     25
80746     24
19348     22
93486     21
23548     12
43658     10
46329      9
76530      8
23457      7
23596      7
98346      3
82963      0
19374      0
92736      0
33496      0
43096      0
98476      0
Name: session_likes_given, dtype: int64

## Insight 10 : 26th October generated 247 likes in total being the highest record for October 2019. The customer with most likes with a total of 230 likes is customer no. 87323

In [46]:
data1.groupby('login_date')['session_comments_given'].sum().sort_values(ascending=False).head(10)

login_date
2019-10-30    49
2019-10-26    48
2019-10-28    42
2019-10-24    42
2019-10-20    35
2019-10-10    34
2019-10-22    33
2019-10-06    32
2019-10-19    32
2019-10-15    31
Name: session_comments_given, dtype: int64

In [47]:
data1.groupby('customer_id')['session_comments_given'].sum().sort_values(ascending=False).head(10)

customer_id
23404    64
40572    46
29375    44
38459    43
87323    41
14354    38
73245    36
40235    34
87265    33
51243    26
Name: session_comments_given, dtype: int64

## Insight 11 : 30th October generated 49 comments in total being the highest record for October 2019. The customer with most comments with a total of 64 comments is customer no. 23404

In [48]:
data1.groupby('customer_id')['inactive_duration'].sum().sort_values(ascending=False).head(10)

customer_id
73245    16527
40572    15302
40235    14931
23404    14630
38459    12554
23083    12030
29375    11867
23985     9290
14354     8964
87323     8728
Name: inactive_duration, dtype: int64

## Insight 12 : Customer no. 73245 has clocked in a total of 16527 seconds of inactivity duration and should be addressed

In [49]:
a1 = data1.groupby(['customer_id'])['session_likes_given'].agg('sum')

In [50]:
cust_likes= pd.DataFrame({'Customer_id':a1.index, 'Total likes given':a1.values})
cust_likes

Customer_id  Total likes given
0         10246                 53
1         12407                 82
2         12496                 53
3         12940                 89
4         14354                150
5         19348                 22
6         19374                  0
7         20548                 25
8         23083                112
9         23404                207
10        23457                  7
11        23548                 12
12        23579                 71
13        23596                  7
14        23985                115
15        24520                 29
16        29345                 27
17        29356                 87
18        29375                189
19        32047                 25
20        33496                  0
21        34574                 61
22        38459                187
23        39475                 47
24        40235                119
25        40572                179
26        43057                 64
27        43086                 45
28        43096                  0
29        43549                 52
30        43658                 10
31        46329                  9
32        51243                154
33        68754                 25
34        73245                183
35        73524                 65
36        76530                  8
37        80746                 24
38        82963                  0
39        87265                149
40        87323                230
41        92736                  0
42        93464                 44
43        93486                 21
44        96254                 59
45        98346                  3
46        98476                  0
47        98653                 27

In [51]:
a2 = data1.groupby(['customer_id'])['session_comments_given'].agg('sum')

In [52]:
cust_comments= pd.DataFrame({'Customer_id':a2.index, 'Total comments given':a2.values})
cust_comments

Customer_id  Total comments given
0         10246                     3
1         12407                    24
2         12496                    23
3         12940                    15
4         14354                    38
5         19348                     1
6         19374                     4
7         20548                     0
8         23083                    14
9         23404                    64
10        23457                     2
11        23548                     0
12        23579                    14
13        23596                     5
14        23985                    12
15        24520                     9
16        29345                     4
17        29356                    16
18        29375                    44
19        32047                    15
20        33496                     0
21        34574                    22
22        38459                    43
23        39475                     8
24        40235                    34
25        40572                    46
26        43057                    24
27        43086                     8
28        43096                     0
29        43549                    20
30        43658                     0
31        46329                     0
32        51243                    26
33        68754                     2
34        73245                    36
35        73524                    11
36        76530                     3
37        80746                    13
38        82963                     0
39        87265                    33
40        87323                    41
41        92736                     5
42        93464                     5
43        93486                     1
44        96254                    16
45        98346                     3
46        98476                     1
47        98653                    14

In [53]:
a3 = data1.groupby(['customer_id'])['session_id'].agg('count')

In [54]:
cust_sessions= pd.DataFrame({'Customer_id':a3.index, 'Total Sessions':a3.values})
cust_sessions

Customer_id  Total Sessions
0         10246               4
1         12407               8
2         12496               6
3         12940               5
4         14354              15
5         19348               1
6         19374               1
7         20548               1
8         23083              11
9         23404              21
10        23457               1
11        23548               1
12        23579               5
13        23596               1
14        23985               8
15        24520               2
16        29345               1
17        29356               6
18        29375              22
19        32047               4
20        33496               1
21        34574               7
22        38459              18
23        39475               4
24        40235              17
25        40572              16
26        43057               6
27        43086               5
28        43096               1
29        43549               7
30        43658               1
31        46329               1
32        51243              12
33        68754               1
34        73245              15
35        73524               8
36        76530               1
37        80746               4
38        82963               1
39        87265              12
40        87323              17
41        92736               1
42        93464               4
43        93486               1
44        96254               7
45        98346               1
46        98476               1
47        98653               6

In [55]:
a4 = data1.groupby(['customer_id'])['bugs_in_session'].agg('sum')

In [57]:
cust_bugs_count= pd.DataFrame({'Customer_id':a4.index, 'Total Bugs Occured':a4.values})
cust_bugs_count

Customer_id  Total Bugs Occured
0         10246                   9
1         12407                   6
2         12496                   4
3         12940                   3
4         14354                  15
5         19348                   0
6         19374                   0
7         20548                   1
8         23083                  16
9         23404                  33
10        23457                   0
11        23548                   0
12        23579                   5
13        23596                   0
14        23985                   8
15        24520                   5
16        29345                   0
17        29356                  11
18        29375                  26
19        32047                   0
20        33496                   0
21        34574                  11
22        38459                  15
23        39475                  14
24        40235                  19
25        40572                  25
26        43057                   8
27        43086                   5
28        43096                   0
29        43549                   7
30        43658                   0
31        46329                   0
32        51243                  14
33        68754                   0
34        73245                  24
35        73524                   9
36        76530                   2
37        80746                   8
38        82963                   0
39        87265                  12
40        87323                  25
41        92736                   4
42        93464                   6
43        93486                   0
44        96254                  10
45        98346                   5
46        98476                   0
47        98653                   5

In [58]:
a5 = data1.groupby(['customer_id'])['session_duration'].agg('sum')

In [59]:
cust_Session_duration= pd.DataFrame({'Customer_id':a5.index, 'Total Session Duration':a5.values})
cust_Session_duration

Customer_id  Total Session Duration
0         10246                    2863
1         12407                   12266
2         12496                    6279
3         12940                    6421
4         14354                   17437
5         19348                    1565
6         19374                    1633
7         20548                     258
8         23083                   12135
9         23404                   24809
10        23457                     207
11        23548                    1327
12        23579                    5718
13        23596                     640
14        23985                   13014
15        24520                    1957
16        29345                    1291
17        29356                    6423
18        29375                   30356
19        32047                    7287
20        33496                    1154
21        34574                    8544
22        38459                   22880
23        39475                    4276
24        40235                   21117
25        40572                   17773
26        43057                    7984
27        43086                    6221
28        43096                    1172
29        43549                    6338
30        43658                     522
31        46329                    1299
32        51243                   12556
33        68754                     822
34        73245                   17375
35        73524                    5448
36        76530                    2091
37        80746                    5671
38        82963                    1869
39        87265                   16487
40        87323                   17954
41        92736                     127
42        93464                    5348
43        93486                     872
44        96254                    8342
45        98346                     103
46        98476                    2051
47        98653                    5747

In [60]:
a6 = data1.groupby(['customer_id'])['inactive_duration'].agg('sum')

In [61]:
cust_inactive_duration= pd.DataFrame({'Customer_id':a6.index, 'Total Inactive Duration':a6.values})
cust_inactive_duration

Customer_id  Total Inactive Duration
0         10246                     2435
1         12407                     6038
2         12496                     4878
3         12940                     3521
4         14354                     8964
5         19348                     2195
6         19374                        0
7         20548                     1801
8         23083                    12030
9         23404                    14630
10        23457                        0
11        23548                        0
12        23579                     2979
13        23596                     2077
14        23985                     9290
15        24520                      590
16        29345                        0
17        29356                     6335
18        29375                    11867
19        32047                     5388
20        33496                        0
21        34574                     1043
22        38459                    12554
23        39475                     3263
24        40235                    14931
25        40572                    15302
26        43057                     4326
27        43086                     2597
28        43096                      936
29        43549                     2461
30        43658                      518
31        46329                      181
32        51243                     7294
33        68754                        0
34        73245                    16527
35        73524                     7103
36        76530                      616
37        80746                     3314
38        82963                     1345
39        87265                     6008
40        87323                     8728
41        92736                      176
42        93464                     7141
43        93486                        0
44        96254                     2681
45        98346                     1248
46        98476                        0
47        98653                     4569

In [62]:
cust_likes = cust_likes.merge(cust_comments, on='Customer_id')
cust_likes = cust_likes.merge(cust_sessions, on='Customer_id')
cust_likes = cust_likes.merge(cust_Session_duration, on='Customer_id')
cust_likes = cust_likes.merge(cust_inactive_duration, on='Customer_id')
cust_likes = cust_likes.merge(cust_bugs_count, on='Customer_id')

In [63]:
Customer_data = cust_likes.copy()

In [64]:
Customer_data

Customer_id  Total likes given  Total comments given  Total Sessions  \
0         10246                 53                     3               4   
1         12407                 82                    24               8   
2         12496                 53                    23               6   
3         12940                 89                    15               5   
4         14354                150                    38              15   
5         19348                 22                     1               1   
6         19374                  0                     4               1   
7         20548                 25                     0               1   
8         23083                112                    14              11   
9         23404                207                    64              21   
10        23457                  7                     2               1   
11        23548                 12                     0               1   
12        23579                 71                    14               5   
13        23596                  7                     5               1   
14        23985                115                    12               8   
15        24520                 29                     9               2   
16        29345                 27                     4               1   
17        29356                 87                    16               6   
18        29375                189                    44              22   
19        32047                 25                    15               4   
20        33496                  0                     0               1   
21        34574                 61                    22               7   
22        38459                187                    43              18   
23        39475                 47                     8               4   
24        40235                119                    34              17   
25        40572                179                    46              16   
26        43057                 64                    24               6   
27        43086                 45                     8               5   
28        43096                  0                     0               1   
29        43549                 52                    20               7   
30        43658                 10                     0               1   
31        46329                  9                     0               1   
32        51243                154                    26              12   
33        68754                 25                     2               1   
34        73245                183                    36              15   
35        73524                 65                    11               8   
36        76530                  8                     3               1   
37        80746                 24                    13               4   
38        82963                  0                     0               1   
39        87265                149                    33              12   
40        87323                230                    41              17   
41        92736                  0                     5               1   
42        93464                 44                     5               4   
43        93486                 21                     1               1   
44        96254                 59                    16               7   
45        98346                  3                     3               1   
46        98476                  0                     1               1   
47        98653                 27                    14               6   

    Total Session Duration  Total Inactive Duration  Total Bugs Occured  
0                     2863                     2435                   9  
1                    12266                     6038                   6  
2                     6279                     4878      

## Insight 13 : This is customer analysis table capturing all metrics for each customer. A strategy can be formulated from this for the poorly performing customers.